lib/dynflow/middleware/common/singleton.rb

```diff
@@ -0,0 +1,24 @@
+module Dynflow
+  module Middleware::Common
+    class Singleton < Middleware
+      def plan(*args)
+        action.singleton_lock!
+        pass(*args)
+        unless action.respond_to?(:run) || action.respond_to?(:finalize)
+          action.singleton_unlock!
+        end
+      end
+
+      def run(*args)
+        action.singleton_lock! unless action.holds_singleton_lock?
+        pass(*args)
+        action.singleton_unlock! unless action.respond_to?(:finalize)
+      end
+
+      def finalize(*args)
```

Comment:
Just capturing our yesterday's conversation: there is `finalize_phase` method, that wraps the whole finalize phase. It hoverer, doesn't expose the action. Therefore, I would vote for keeping it this way.

Reply:
Meanwhile I somehow managed to get it working using finalize_phase. Explanatory comment incoming